README.md
# adaptive-scroll v0.2.0 [![Build Status](https://travis-ci.org/angular-adaptive/adaptive-scroll.png?branch=master)](https://travis-ci.org/angular-adaptive/adaptive-scroll)

This module allows you to scroll an AngularJS app using gyroscope.

**Note: You need a device with a gyroscope and a browser with requestAnimationFrame support for smooth scrolling**

**We recommend [Gyrocopter extension for Chrome](https://github.com/janantala/Gyrocopter).**

### Demo

<p align="center">
    Check out http://angular-adaptive.github.io/adaptive-scroll/demo/ Demo image with Gyrocopter extension.
</p>
<p align="center">
    <img src="https://raw.github.com/angular-adaptive/adaptive-scroll/master/images/gyrocopter.gif" alt="demo" />
</p>
# Requirements

- AngularJS v 1.0+
- device with a gyroscope or a [simulator](https://github.com/janantala/Gyrocopter)
- browser with requestAnimationFrame support for smooth scrolling 
    - iOS Safari 6+, 
    - Chrome for Android 25+, 
    - Opera Mobile 14+
    - Firefox for Android 19+

# Usage

We use [bower](http://twitter.github.com/bower/) for dependency management. Add

    dependencies: {
        "angular-adaptive-scroll": "latest"
    }

To your `bower.json` file. Then run

    bower install

This will copy the scroll files into your `bower_components` folder, along with its dependencies. Load the script files in your application:

    <script type="text/javascript" src="bower_components/angular/angular.js"></script>
    <script type="text/javascript" src="bower_components/angular-adaptive-scroll/angular-adaptive-scroll.min.js"></script>

**Add the adaptive.scroll module** as a dependency to your application module:

    var myAppModule = angular.module('MyApp', ['adaptive.scroll']);

and **include $gyroscope service** as a dependency to your controller:

    angular.module('MyApp').controller('MainCtrl', function ['$scope', '$gyroscope', ($scope, $gyroscope) {

    }]);

**Append a directive adaptivescroll** to the scrollable element.

    <body adaptivescroll>
      ...
    </body>
    
    or
    
    <textarea adaptivescroll></textarea>

To start scrolling **run method watchPosition()** and pass trashold argument (in degrees):

    $gyroscope.watchPosition(10);

To stop scrolling run method ignorePosition():

    $gyroscope.ignorePosition();

There are also dirrection events returning direction diff you can handle

    $gyroscope.onalpha(function(alphaDiff){
    	console.log(alphaDiff);
	});
	$gyroscope.onbeta(function(betaDiff){
		console.log(betaDiff);
	});
	$gyroscope.ongamma(function(gammaDiff){
		console.log(gammaDiff);
	});

# Testing

We use karma and jshint to ensure the quality of the code. The easiest way to run these checks is to use grunt:

    npm install -g grunt-cli
    npm install
    bower install
    grunt

The karma task will try to open Chrome as a browser in which to run the tests. Make sure this is available or change the configuration in `test/test.config.js` 

# Contributing

Pull requests are welcome. 

Make a PR against canary branch and don't bump any versions. 

Please respect the code style in place.

# License

The MIT License

Copyright (c) 2013 Jan Antala, https://github.com/janantala


CHANGELOG.md
<a name="v0.2.0"></a>
## v0.2.0 (2013-11-12)


#### Features

* **GruntFile:** add concated and uglified version ([cb8aab2d](https://github.com/angular-adaptive/adaptive-scroll/commit/cb8aab2d6a38c6f8464b56264e313cf368120295))
* **conventional-changelog:** adds grunt-conventional-changelog support ([f11e9a41](https://github.com/angular-adaptive/adaptive-scroll/commit/f11e9a410ed49d3c728cbe2d9cd1f8e43771bdf4))


#### Breaking Changes

* src/adaptive-scroll.js is now renamed into src/angular-adaptive-scroll.js
 ([cb8aab2d](https://github.com/angular-adaptive/adaptive-scroll/commit/cb8aab2d6a38c6f8464b56264e313cf368120295))



test/test.conf.js
module.exports = function(config) {
  config.set({
    basePath: '..',
    files: [
      'components/angular/angular.js',
      'components/angular-mocks/angular-mocks.js',
      'src/angular-adaptive-scroll.js',
      'test/*.spec.js'
    ],
    frameworks: ['jasmine'],
    singleRun: true,
    browsers: [ 'Chrome' ]
  });
};

test/unit.spec.js
describe('adaptive.scroll', function() {

  var rootscope;

  beforeEach(module('adaptive.scroll'));

  beforeEach(inject(function($rootScope) {
    rootScope = $rootScope;
    spyOn(rootScope, "$on").andCallThrough();
    spyOn(rootScope, "$broadcast").andCallThrough();
  }));

  describe('$gyroscope service', function() {

    var $gyroscope;

    beforeEach(inject(function (_$gyroscope_) {
      $gyroscope = _$gyroscope_;
    }));

    it('should be an object', function () {
      expect(typeof $gyroscope).toBe('object');
    });

    it('should have methods watchPosition(), ignorePosition()', function () {
      expect($gyroscope.watchPosition).toBeDefined();
      expect($gyroscope.ignorePosition).toBeDefined();

      expect(typeof $gyroscope.watchPosition).toBe('function');
      expect(typeof $gyroscope.ignorePosition).toBe('function');
    });

  });


  describe('adaptivescroll directive', function() {
    var elm, scope;

    beforeEach(inject(function($rootScope, $compile) {
      elm = angular.element(
        '<textarea adaptivescroll>' +
        '</textarea>'
      );

      scope = $rootScope;

      $compile(elm)(scope);
      scope.$digest();
    }));

    it('should have called rootScope.$on', function(){
      expect(rootScope.$on).toHaveBeenCalled();
    });


  });


});

Gruntfile.js
module.exports = function (grunt) {
    require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

    // Default task.
    grunt.registerTask('default', ['karma', 'jshint']);

    var karmaConfig = function(configFile, customOptions) {
        var options = { configFile: configFile, keepalive: true };
        var travisOptions = process.env.TRAVIS && { browsers: ['Firefox'], reporters: 'dots' };
        return grunt.util._.extend(options, customOptions, travisOptions);
    };

    // Project configuration.
    grunt.initConfig({
        karma: {
            unit: {
            options: karmaConfig('test/test.conf.js')
            }
        },
        jshint:{
            files:['src/**/*.js', 'test/**/*.js', 'demo/js/controllers/mainCtrl.js'],
            options: {
                curly:true,
                eqeqeq:true,
                immed:true,
                latedef:true,
                newcap:true,
                noarg:true,
                sub:true,
                boss:true,
                eqnull:true,
                devel:true,
                globals:{}
            }
        },
        changelog: {
          options: {
            dest: 'CHANGELOG.md'
          }
        }
    });
}


angular-adaptive-scroll.js
/*!
 * angular-adaptive-scroll v0.2.0
 * The MIT License
 * Copyright (c) 2013 Jan Antala
 */

(function () {

var adaptive = angular.module('adaptive.scroll', []);

adaptive.factory('$gyroscope', ['$rootScope', function ($rootScope) {

  var alphaStart, betaStart, gammaStart;
  var alpha, beta, gamma;
  var trashold = 20;
  var active = false;
  var onalpha;
  var onbeta;
  var ongamma;

  window.ondeviceorientation = function(event) {
    if (!active) {
      return false;
    }

    alpha = Math.round(event.alpha);
    beta = Math.round(event.beta);
    gamma = Math.round(event.gamma);

    if (!alphaStart && !betaStart && !gammaStart) {
      alphaStart = alpha;
      betaStart = beta;
      gammaStart = gamma;
    }

    var alphaDiff = getDiff(alphaStart, alpha, -360);
    var betaDiff = getDiff(betaStart, beta, 360);
    var gammaDiff = getDiff(gammaStart, gamma, 180);

    if ((Math.abs(alphaDiff) <= trashold) && (Math.abs(betaDiff) <= trashold) && (Math.abs(gammaDiff) <= trashold)) {
      stopEvent();
    }
    else {
      sendEvent(alphaDiff, betaDiff, gammaDiff);
    }
  };

  var getDiff = function(a, b, interval){
    d1 = a - b;
    d2 = a - b - interval;
    return (Math.abs(d1) < Math.abs(d2)) ? d1 : d2;
  };

  var sendEvent = function(alphaDiff, betaDiff, gammaDiff) {

    if (Math.abs(alphaDiff) > trashold) {
      alphaDiff = alphaDiff < 0 ? alphaDiff + trashold : alphaDiff - trashold;
    }
    if (Math.abs(betaDiff) > trashold) {
      betaDiff = betaDiff < 0 ? betaDiff + trashold : betaDiff - trashold;
    }
    if (Math.abs(gammaDiff) > trashold) {
      gammaDiff = gammaDiff < 0 ? gammaDiff + trashold : gammaDiff - trashold;
    }

    $rootScope.$broadcast('adaptive.scroll:deviceorientation', {'event': 'ondeviceorientation', 'alphaDiff': alphaDiff, 'betaDiff': betaDiff, 'gammaDiff': gammaDiff});

    if (onalpha && alphaDiff !== 0){
      onalpha(alphaDiff);
    }
    if (onbeta && betaDiff !== 0){
      onbeta(betaDiff);
    }
    if (ongamma && gammaDiff !== 0){
      ongamma(gammaDiff);
    }
  };

  var startEvent = function() {
        $rootScope.$broadcast('adaptive.scroll:deviceorientation', {'event': 'ondeviceorientationstart'});
  };

  var stopEvent = function() {
        $rootScope.$broadcast('adaptive.scroll:deviceorientation', {'event': 'ondeviceorientationstop'});
  };

  var watchPosition = function(degrees) {
    console.log('trashold', degrees);
    trashold = degrees || trashold;
    active = true;
  };

  var ignorePosition = function() {
    active = false;
    alphaStart = null;
    betaStart = null;
    gammaStart = null;
    stopEvent();
  };

  return {
    watchPosition: function(degrees) {
      watchPosition(degrees);
    },
    ignorePosition: function() {
      ignorePosition();
    },
    onalpha: function(fn) {
      onalpha = fn;
    },
    onbeta: function(fn) {
      onbeta = fn;
    },
    ongamma: function(fn) {
      ongamma = fn;
    }
  };

}]);

adaptive.directive('adaptivescroll', ['$rootScope', function ($rootScope) {
  return {
    restrict: 'A',
    link: function (scope, element, attrs) {
      var getOptions = function () {
        return angular.extend({}, scope.$eval(attrs.adaptivescroll));
      };
      var opts = getOptions();
      console.log(opts);


      (function() {
        var lastTime = 0;
        var vendors = ['webkit', 'moz'];
        for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
            window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
            window.cancelAnimationFrame =
              window[vendors[x]+'CancelAnimationFrame'] || window[vendors[x]+'CancelRequestAnimationFrame'];
        }

        if (!window.requestAnimationFrame) {
            window.requestAnimationFrame = function(callback, element) {
                var currTime = new Date().getTime();
                var timeToCall = Math.max(0, 16 - (currTime - lastTime));
                var id = window.setTimeout(function() { callback(currTime + timeToCall); },
                  timeToCall);
                lastTime = currTime + timeToCall;
                return id;
            };
        }

        if (!window.cancelAnimationFrame) {
            window.cancelAnimationFrame = function(id) {
                clearTimeout(id);
            };
        }
      }());

      var alphaDiff = 0;
      var betaDiff = 0;
      var gammaDiff = 0;
      var requestId;

      var scroll = function() {
        if (alphaDiff || betaDiff) {
          element[0].scrollTop = element[0].scrollTop + betaDiff;
          element[0].scrollLeft = element[0].scrollLeft + gammaDiff;
        }
        requestId = window.requestAnimationFrame(scroll);
      };

      $rootScope.$on('adaptive.scroll:deviceorientation', function(e, data){

        console.log('received', data);

        if (data.event === 'ondeviceorientation') {
          alphaDiff = data.alphaDiff;
          betaDiff = data.betaDiff;
          gammaDiff = data.gammaDiff;
        }

        if (data.event === 'ondeviceorientation' && !requestId) {
          requestId = window.requestAnimationFrame(scroll);
        }

        if (data.event === 'ondeviceorientationstop' && requestId) {
          window.cancelAnimationFrame(requestId);
          requestId = null;
        }

      });

    }
  };
}]);

})();


src/adaptive-scroll.js
(function () {
/**
 * @ngdoc overview
 * @name adaptive.scroll
 *
 * @description
 * The `adaptive.scroll` module provides components to realize adaptive scrolling
 * in your Angular applications. You can use its components to, for example,
 * let the user scroll in your app by changing the alpha, beta and gamma relations
 * on his mobile device.
 */
var adaptive = angular.module('adaptive.scroll', []);

/**
 * @ngdoc object
 * @name adaptive.scroll.$gyroscope
 * @requires $rootScope
 *
 * @description
 * The `$gyroscope` is your interface to communicate with the device' gyroscope.
 * It provides methods to react on alpha, beta or gamma value changes and it also
 * let's you scroll through your application programatically.
 */
adaptive.factory('$gyroscope', ['$rootScope', function ($rootScope) {

  var alphaStart, betaStart, gammaStart;
  var alpha, beta, gamma;
  var trashold = 20;
  var active = false;
  var onalpha;
  var onbeta;

  var ongamma;

  window.ondeviceorientation = function(event) {
    if (!active) {
      return false;
    }

    alpha = Math.round(event.alpha);
    beta = Math.round(event.beta);
    gamma = Math.round(event.gamma);

    if (!alphaStart && !betaStart && !gammaStart) {
      alphaStart = alpha;
      betaStart = beta;
      gammaStart = gamma;
    }

    var alphaDiff = getDiff(alphaStart, alpha, -360);
    var betaDiff = getDiff(betaStart, beta, 360);
    var gammaDiff = getDiff(gammaStart, gamma, 180);

    if ((Math.abs(alphaDiff) <= trashold) && (Math.abs(betaDiff) <= trashold) && (Math.abs(gammaDiff) <= trashold)) {
      stopEvent();
    }
    else {
      sendEvent(alphaDiff, betaDiff, gammaDiff);
    }
  };

  var getDiff = function(a, b, interval){
    d1 = a - b;
    d2 = a - b - interval;
    return (Math.abs(d1) < Math.abs(d2)) ? d1 : d2;
  };

  var sendEvent = function(alphaDiff, betaDiff, gammaDiff) {

    if (Math.abs(alphaDiff) > trashold) {
      alphaDiff = alphaDiff < 0 ? alphaDiff + trashold : alphaDiff - trashold;
    }
    if (Math.abs(betaDiff) > trashold) {
      betaDiff = betaDiff < 0 ? betaDiff + trashold : betaDiff - trashold;
    }
    if (Math.abs(gammaDiff) > trashold) {
      gammaDiff = gammaDiff < 0 ? gammaDiff + trashold : gammaDiff - trashold;
    }

    $rootScope.$broadcast('adaptive.scroll:deviceorientation', {'event': 'ondeviceorientation', 'alphaDiff': alphaDiff, 'betaDiff': betaDiff, 'gammaDiff': gammaDiff});

    if (onalpha && alphaDiff !== 0){
      onalpha(alphaDiff);
    }
    if (onbeta && betaDiff !== 0){
      onbeta(betaDiff);
    }
    if (ongamma && gammaDiff !== 0){
      ongamma(gammaDiff);
    }
  };

  var startEvent = function() {
        $rootScope.$broadcast('adaptive.scroll:deviceorientation', {'event': 'ondeviceorientationstart'});
  };

  var stopEvent = function() {
        $rootScope.$broadcast('adaptive.scroll:deviceorientation', {'event': 'ondeviceorientationstop'});
  };

  /**
   * @ngdoc function
   * @name adaptive.scroll.$gyroscope#watchPosition
   * @methodOf adaptive.scope.$gyroscope
   *
   * @description
   * This method starts scrolling by considering the given trashold in degrees.
   *
   * @param {number} degress Trashold in degrees.
   */
  var watchPosition = function(degrees) {
    console.log('trashold', degrees);
    trashold = degrees || trashold;
    active = true;
  };

  /**
   * @ngdoc function
   * @name adaptive.scroll.$gyroscope#ignorePosition
   * @method adaptive.scroll.$gyroscope
   *
   * @description
   * Stops current scrolling.
   */
  var ignorePosition = function() {
    active = false;
    alphaStart = null;
    betaStart = null;
    gammaStart = null;
    stopEvent();
  };

  return {
    watchPosition: function(degrees) {
      watchPosition(degrees);
    },
    ignorePosition: function() {
      ignorePosition();
    },
    /**
     * @ngdoc function
     * @name adaptive.scroll.$gyroscope#onalpha
     * @method adaptive.scroll.$gyroscope
     *
     * @description
     * Registers callback which gets executed when alpha direction changes.
     * You can access the alpha value diff in the callback.
     *
     * @param {object} fn Callback function
     */
    onalpha: function(fn) {
      onalpha = fn;
    },
    /**
     * @ngdoc function
     * @name adaptive.scroll.$gyroscope#onbeta
     * @method adaptive.scroll.$gyroscope
     *
     * @description
     * Registers callback which gets executed when beta direction changes.
     * You can access the beta value diff in the callback.
     *
     * @param {object} fn Callback function
     */
    onbeta: function(fn) {
      onbeta = fn;
    },
    /**
     * @ngdoc function
     * @name adaptive.scroll.$gyroscope#ongamma
     * @method adaptive.scroll.$gyroscope
     *
     * @description
     * Registers callback which gets executed when gamma direction changes.
     * You can access the gamma value diff in the callback.
     *
     * @param {object} fn Callback function
     */
    ongamma: function(fn) {
      ongamma = fn;
    }
  };

}]);

/**
 * @ngdoc object
 * @name adaptive.scroll.directive:adaptivescroll
 * @requires $rootScope
 * @restrict A
 *
 * @description
 * Apply this directive to any scrollable element to make it adaptively scrollable.
 */
adaptive.directive('adaptivescroll', ['$rootScope', function ($rootScope) {
  return {
    restrict: 'A',
    link: function (scope, element, attrs) {
      var getOptions = function () {
        return angular.extend({}, scope.$eval(attrs.adaptivescroll));
      };
      var opts = getOptions();
      console.log(opts);


      (function() {
        var lastTime = 0;
        var vendors = ['webkit', 'moz'];
        for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
            window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
            window.cancelAnimationFrame =
              window[vendors[x]+'CancelAnimationFrame'] || window[vendors[x]+'CancelRequestAnimationFrame'];
        }

        if (!window.requestAnimationFrame) {
            window.requestAnimationFrame = function(callback, element) {
                var currTime = new Date().getTime();
                var timeToCall = Math.max(0, 16 - (currTime - lastTime));
                var id = window.setTimeout(function() { callback(currTime + timeToCall); },
                  timeToCall);
                lastTime = currTime + timeToCall;
                return id;
            };
        }

        if (!window.cancelAnimationFrame) {
            window.cancelAnimationFrame = function(id) {
                clearTimeout(id);
            };
        }
      }());

      var alphaDiff = 0;
      var betaDiff = 0;
      var gammaDiff = 0;
      var requestId;

      var scroll = function() {
        if (alphaDiff || betaDiff) {
          element[0].scrollTop = element[0].scrollTop + betaDiff;
          element[0].scrollLeft = element[0].scrollLeft + gammaDiff;
        }
        requestId = window.requestAnimationFrame(scroll);
      };

      $rootScope.$on('adaptive.scroll:deviceorientation', function(e, data){

        console.log('received', data);

        if (data.event === 'ondeviceorientation') {
          alphaDiff = data.alphaDiff;
          betaDiff = data.betaDiff;
          gammaDiff = data.gammaDiff;
        }

        if (data.event === 'ondeviceorientation' && !requestId) {
          requestId = window.requestAnimationFrame(scroll);
        }

        if (data.event === 'ondeviceorientationstop' && requestId) {
          window.cancelAnimationFrame(requestId);
          requestId = null;
        }

      });

    }
  };
}]);

})();
